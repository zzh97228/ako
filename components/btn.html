<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>
      Btn | Acco
    </title>
    <meta name="description" content="Documentation of acco">
    <link rel="stylesheet" href="/acco/assets/style.cea082f4.css">
    <link rel="modulepreload" href="/acco/assets/framework.27075354.js">
    <link rel="modulepreload" href="/acco/assets/components_btn.md.8f217747.lean.js">
    <link rel="modulepreload" href="/acco/assets/app.942d5e25.js">
    
    
  </head>
  <body>
    <div id="app"><!----></div>
    <script>__VP_HASH_MAP__ = JSON.parse("{\"index.md\":\"d54369d9\",\"components_btn.md\":\"8f217747\",\"components_card.md\":\"b3d08341\",\"components_carousel.md\":\"36a145cf\",\"components_checkbox.md\":\"bc00845b\",\"components_field.md\":\"a57a1d00\",\"components_fileInput.md\":\"bd6e4c36\",\"components_grid.md\":\"084a7e04\",\"components_textBox.md\":\"22918cbc\",\"components_textInput.md\":\"cf4f9f31\",\"directives_ripple.md\":\"a7b455b8\"}")</script>
    <script type="module" async src="/acco/assets/app.942d5e25.js"></script>
  </body>
</html>