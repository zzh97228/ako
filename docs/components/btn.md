# Btn

:tada:

<ac-btn>button</ac-btn>

<ac-btn link>button</ac-btn>

<ac-btn color="primary" outlined>button</ac-btn>

<ac-btn color="primary" flat>button</ac-btn>

<ac-btn tile>button</ac-btn>

### Props

---

| Name        | Type                | Default     | Description                               |
| :---------- | :------------------ | :---------- | :---------------------------------------- |
| tag         | String              | button      | element tag                               |
| outliend    | Boolean             | -           | outlined style                            |
| flat        | Boolean             | -           | flat style                                |
| tile        | Boolean             | -           | square style                              |
| link        | Boolean             | -           | link style                                |
| small       | Boolean             | -           | small style                               |
| large       | Boolean             | -           | large style                               |
| block       | Boolean             | -           | display block and width 100%              |
| round       | Boolean             | -           | round style                               |
| disabled    | Boolean             | -           | disabled to click                         |
| icon        | Boolean             | -           | limit width and height to be equal        |
| loading     | \[Boolean, String\] | -           | loading type button and disabled to click |
| color       | String              | primary     | change current color class or style       |
| elevation   | String              | -           | change current shadow class or style      |
| toggleable  | Boolean             | -           | set this element whether to be toggleable |
| active      | Boolean             | -           | control inside active `ref` from outside  |
| activeClass | String              | btn--active | active class name                         |

### Loading

---

<ac-row column gutter="xs">
<ac-col><ac-btn loading="Loading..."></ac-btn></ac-col>
<ac-col><ac-btn icon loading>èµž</ac-btn></ac-col>
</ac-row>
