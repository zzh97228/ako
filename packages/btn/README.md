# `<btn>`

## Props

| Name        | Type                | Default     | Description                               |
| :---------- | :------------------ | :---------- | :---------------------------------------- |
| tag         | String              | button      | element tag                               |
| outliend    | Boolean             | -           | outlined style                            |
| flat        | Boolean             | -           | flat style                                |
| tile        | Boolean             | -           | square style                              |
| link        | Boolean             | -           | link style                                |
| small       | Boolean             | -           | small style                               |
| large       | Boolean             | -           | large style                               |
| block       | Boolean             | -           | display block and width 100%              |
| round       | Boolean             | -           | round style                               |
| disabled    | Boolean             | -           | disabled to click                         |
| icon        | Boolean             | -           | limit width and height to be equal        |
| loading     | \[Boolean, String\] | -           | loading type button and disabled to click |
| color       | String              | primary     | change current color class or style       |
| elevation   | String              | -           | change current shadow class or style      |
| toggleable  | Boolean             | -           | set this element whether to be toggleable |
| active      | Boolean             | -           | control inside active `ref` from outside  |
| activeClass | String              | btn--active | active class name                         |
