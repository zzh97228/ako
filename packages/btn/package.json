{
  "name": "@lagabu/btn",
  "main": "dist/btn.umd.bundle.js",
  "types": "dist/index.d.ts",
  "module": "dist/ako-ui.es.bundle.js",
  "version": "0.1.0-alpha.0",
  "files": [
    "dist",
    "README.md",
    "dist/index.d.ts"
  ],
  "dependencies": {
    "@lagabu/shared": "latest",
    "vue": "next"
  },
  "peerDependencies": {
    "vue": "next"
  }
}