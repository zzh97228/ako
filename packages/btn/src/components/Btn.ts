import vue, { VNodeArrayChildren } from 'vue';
import {
  Ripple,
  useColor,
  useElevation,
  genElevationProp,
  genColorProp,
  genToggleProps,
  useToggle,
  isString,
} from '@lagabu/shared';
import { useGroupConsumer } from '../composables';
import { computed, defineComponent, h, mergeProps, withDirectives } from 'vue';
export default defineComponent({
  name: 'btn',
  props: {
    tag: {
      type: String,
      default: 'button',
    },
    outlined: Boolean,
    flat: Boolean,
    tile: Boolean,
    small: Boolean,
    large: Boolean,
    block: Boolean,
    round: Boolean,
    disabled: Boolean,
    link: Boolean,
    icon: Boolean,
    loading: [Boolean, String],
    ...genColorProp(),
    ...genElevationProp(),
    ...genToggleProps('btn--active'),
  },
  setup(props, context) {
    const isTextColor = computed(() => props.outlined || props.link);
    const color = useColor(props, isTextColor);
    const elevation = useElevation(props);
    const { isActive, class: toggleClasses, toggle } = useToggle(props, context);
    const group = useGroupConsumer(props, isActive);
    const notAllowed = computed(() => props.loading || props.disabled);
    const classes = computed(() => {
      return {
        btn: true,
        'btn--small': props.small,
        'btn--large': props.large,
        'btn--tile': props.tile,
        'btn--flat': props.flat,
        'btn--outlined': props.outlined,
        'btn--block': props.block,
        'btn--round': props.round,
        'btn--disabled': notAllowed.value,
        'btn--link': props.link,
        'btn--icon': props.icon,
      };
    });
    return {
      classes,
      colorClasses: color.class,
      colorStyles: color.style,
      elevationClasses: elevation.class,
      elevationStyles: elevation.style,
      toggleClasses,
      notAllowed,
      onClick: (e: MouseEvent) => {
        if (notAllowed.value) {
          e.preventDefault();
          e.stopImmediatePropagation();
          return;
        }
        toggle(group.onToggle, group.notAllowed);
      },
    };
  },
  methods: {
    genSpin() {
      return h('div', {
        class: 'btn__spinner',
      });
    },
    genLoading() {
      return h(
        'span',
        {
          class: 'btn__loading',
        },
        isString(this.loading) && this.loading ? this.loading : 'loading...'
      );
    },
    genLoadingSlot() {
      const vnodeArr: VNodeArrayChildren = [this.genSpin()];
      if (!this.icon) {
        vnodeArr.push(this.genLoading());
      }
      return vnodeArr;
    },
    genBtnContent(children: VNodeArrayChildren | undefined) {
      return h(
        'div',
        {
          class: 'btn__content',
        },
        children ? children : this.$slots.default && this.$slots.default()
      );
    },
    genBtn() {
      return h(
        this.tag,
        mergeProps(
          {
            class: this.classes,
            onClick: this.onClick,
          },
          {
            class: this.colorClasses,
            style: this.colorStyles,
          },
          {
            class: this.elevationClasses,
            style: this.elevationStyles,
          },
          {
            class: this.toggleClasses,
          }
        ),
        this.genBtnContent(this.loading ? this.genLoadingSlot() : void 0)
      );
    },
  },
  render() {
    return withDirectives(this.genBtn(), [
      [
        Ripple,
        {
          disabled: this.notAllowed,
        },
        this.tile ? 'tile' : '',
      ],
    ]);
  },
});
