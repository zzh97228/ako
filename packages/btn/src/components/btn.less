.btn {
  @normal-size: 33px;
  position: relative;
  text-decoration: none;
  user-select: none;
  outline: none;
  border: 1px solid transparent;
  border-radius: 4px;
  padding: 5px 11px;
  font-size: 14px;
  height: @normal-size;
  min-width: @normal-size;
  color: white;
  cursor: pointer;
  vertical-align: middle;
  opacity: 1;
  transition: all 300ms ease;

  &__content {
    overflow: hidden;
    line-height: 1.5;
  }

  &:not(.btn--disabled):hover {
    opacity: 0.85;
  }

  &--outlined {
    border-color: currentColor;
    background-color: transparent;
  }

  &--flat,
  &--outlined {
    box-shadow: none;
  }

  &--tile {
    border-radius: 0;
  }

  @small-size: 29px;
  @large-size: 41px;

  &--small {
    padding: 3px 11px;
    height: @small-size;
    min-width: @small-size;
  }

  &--large {
    padding: 9px 11px;
    height: @large-size;
    min-width: @large-size;
  }

  &--round {
    border-radius: 50%;
  }

  &--round,
  &--fab {
    padding: 5px;
    width: @normal-size;

    &.btn--large {
      padding: 9px;
      width: @large-size;
    }

    &.btn--small {
      padding: 3px;
      width: @small-size;
    }
  }

  &--block {
    width: 100%;
    max-width: 100%;
    display: block;
  }

  &--link {
    color: var(--black, black);
    border-color: transparent !important;
    background-color: transparent !important;
  }

  &:active {
    box-shadow: none;
  }

  &--disabled {
    cursor: not-allowed;
    background-color: var(--black-lighten) !important;
    border-color: var(--black-lighten) !important;
    box-shadow: none !important;
  }
}
