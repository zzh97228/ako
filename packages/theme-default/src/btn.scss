@use './variables/index' as v;
@use 'sass:map';

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

.btn {
  $normal-padding: map.get(v.$vertical-padding, 'md');
  $large-padding: map.get(v.$vertical-padding, 'lg');
  $small-padding: map.get(v.$vertical-padding, 'sm');
  $justify-padding: v.$justify-padding;
  $min-height: v.$inline-font-size * v.$inline-line-height + v.$border-size * 2 + $normal-padding * 2;
  $min-width: $min-height;
  display: inline-block;
  vertical-align: middle;
  outline: none;
  cursor: pointer;
  border: v.$border-size solid transparent;
  border-radius: 4px;
  min-height: $min-height;
  min-width: $min-width;
  padding: $normal-padding $justify-padding;
  font-size: v.$inline-font-size;
  overflow: hidden;
  text-decoration: none;
  position: relative;
  color: var(--grey-0, white);
  background: transparent;
  &:link {
    color: currentColor;
  }
  &__content {
    line-height: v.$inline-line-height;
    text-overflow: ellipsis;
    vertical-align: middle;
  }

  &__loading {
    letter-spacing: 2px;
    vertical-align: middle;
  }

  &__spinner {
    width: 14px;
    height: 14px;
    border-radius: 50%;
    border: 2px solid transparent;
    border-top-color: currentColor;
    display: inline-block;
    animation: spin 1s infinite ease;
    vertical-align: middle;

    & + .btn__loading {
      margin-left: 4px;
    }
  }

  &--small {
    padding: $small-padding $justify-padding;
  }

  &--large {
    font-size: v.$larger-font-size;
    padding: $large-padding $justify-padding;
  }

  &--round,
  &--icon {
    padding: $normal-padding;
    overflow: hidden;
    &.btn--large {
      padding: $large-padding;
    }

    &.btn--small {
      padding: $small-padding;
    }
  }

  &--round {
    border-radius: 50%;
  }

  &--flat,
  &--outlined,
  &--link,
  &--disabled {
    box-shadow: none !important;
  }

  &--link {
    border-radius: 0;
    &::after {
      position: absolute;
      content: '';
      bottom: 0;
      left: 0;
      right: 0;
      height: 2px;
      background-color: currentColor;
      opacity: 0;
      transition: opacity 300ms ease;
    }
    &.btn--active::after {
      opacity: 1;
    }
  }

  &--tile {
    border-radius: 0;
  }
  &--outlined {
    border-color: currentColor;
  }

  &--block {
    display: block;
    width: 100%;
  }

  &--disabled {
    background-color: var(--grey-3, grey);
    cursor: not-allowed;
  }
}
