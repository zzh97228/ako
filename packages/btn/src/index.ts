import './components/btn.less';
export * from './components';
export * from './install';