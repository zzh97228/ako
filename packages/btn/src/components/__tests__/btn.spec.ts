import { mount, VueWrapper } from '@vue/test-utils';
import { defineComponent, h } from 'vue';
import { Btn } from '../btn';

describe('btn.ts', () => {
  let mountFunc: (opts?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (opts = {}) => {
      return mount(Btn, opts);
    };
  });
  it('should render btn with default props', async () => {
    const wrapper = mountFunc({
      props: {
        outlined: true,
        color: 'primary',
        flat: true,
        small: true,
        large: true,
        block: true,
        round: true,
      },
    });

    expect(wrapper.classes()).toContain('elevation-xs');
    expect(wrapper.classes()).toContain('btn');
    expect(wrapper.classes()).toContain('btn--outlined');
    expect(wrapper.classes()).toContain('btn--flat');
    expect(wrapper.classes()).toContain('primary-color--text');
    expect(wrapper.classes()).toContain('btn--small');
    expect(wrapper.classes()).toContain('btn--large');
    expect(wrapper.classes()).toContain('btn--block');
    expect(wrapper.classes()).toContain('btn--round');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not be called when set disabled props', async () => {
    const click = jest.fn();
    const wrapper = mount(() =>
      h(
        Btn,
        {
          disabled: true,
          onClick: click,
        },
        () => 'button'
      )
    );
    await wrapper.trigger('click');
    expect(click).not.toHaveBeenCalled();
  });
});
