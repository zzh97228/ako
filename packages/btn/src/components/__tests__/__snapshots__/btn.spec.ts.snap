// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`btn.ts should render btn with default props 1`] = `
<button class="btn btn--small btn--large btn--flat btn--outlined btn--block btn--round btn--icon primary-color--text elevation-xs">
  <div class="btn__content">
  </div>
</button>
`;
